### ISSUE TITLE
ICE: assertion failed: rows.iter().all(|r| r.len() == v.len())

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6254]
from `./src/test/ui/pattern/usefulness/consts-opaque.rs`
### Code

```rust
#[derive(PartialEq)]
struct Foo(i32);
const FOO_REF_REF: &&Foo = &&Foo(42);

fn main() {
    // This used to cause an ICE (https://github.com/rust-lang/rust/issues/78071)
    match FOO_REF_REF {
        FOO_REF_REF => {}
        Foo(_) => {}
    }
}
```

### Meta

  ```
clippy 0.0.212 (ffa2e7a 2020-10-24)

rustc 1.49.0-nightly (ffa2e7ae8 2020-10-24)
binary: rustc
commit-hash: ffa2e7ae8fbf9badc035740db949b9dae271c29f
commit-date: 2020-10-24
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
LLVM version: 11.0
  ```

### Error output

```
warning: to use a constant of type `Foo` in a pattern, `Foo` must be annotated with `#[derive(PartialEq, Eq)]`
 --> ./src/test/ui/pattern/usefulness/consts-opaque.rs:8:9
  |
8 |         FOO_REF_REF => {}
  |         ^^^^^^^^^^^
  |
  = note: `#[warn(indirect_structural_match)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #62411 <https://github.com/rust-lang/rust/issues/62411>

thread 'rustc' panicked at 'assertion failed: rows.iter().all(|r| r.len() == v.len())', compiler/rustc_mir_build/src/thir/pattern/_match.rs:2030:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.0.212 (ffa2e7a 2020-10-24)

query stack during panic:
#0 [check_match] match-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
warning: 1 warning emitted
```

<details><summary>Backtrace</summary>
  <p>
  
  ```
warning: to use a constant of type `Foo` in a pattern, `Foo` must be annotated with `#[derive(PartialEq, Eq)]`
 --> ./src/test/ui/pattern/usefulness/consts-opaque.rs:8:9
  |
8 |         FOO_REF_REF => {}
  |         ^^^^^^^^^^^
  |
  = note: `#[warn(indirect_structural_match)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #62411 <https://github.com/rust-lang/rust/issues/62411>

thread 'rustc' panicked at 'assertion failed: rows.iter().all(|r| r.len() == v.len())', compiler/rustc_mir_build/src/thir/pattern/_match.rs:2030:5
stack backtrace:
   0:     0x7f29846fba20 - std::backtrace_rs::backtrace::libunwind::trace::h303a626d53553a64
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:100:5
   1:     0x7f29846fba20 - std::backtrace_rs::backtrace::trace_unsynchronized::h6abb882733c3f18c
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f29846fba20 - std::sys_common::backtrace::_print_fmt::h57829321f1d9217a
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f29846fba20 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h400b66a054640aac
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f298476ae9c - core::fmt::write::h1857a60b204f1b6a
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/core/src/fmt/mod.rs:1076:17
   5:     0x7f29846ed642 - std::io::Write::write_fmt::h6453b28c8b27c6ea
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/io/mod.rs:1516:15
   6:     0x7f29847009d5 - std::sys_common::backtrace::_print::h15cc4b3923cc9bc0
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f29847009d5 - std::sys_common::backtrace::print::h953c7be178c0674d
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f29847009d5 - std::panicking::default_hook::{{closure}}::h635d667f571198f0
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:208:50
   9:     0x7f2984700678 - std::panicking::default_hook::h3bc2efb48060abaa
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:227:9
  10:     0x55b778c26491 - clippy_driver::report_clippy_ice::h76ee1ad508e689ae
  11:     0x7f2984701216 - std::panicking::rust_panic_with_hook::h887eb0c89f46b8b2
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:581:17
  12:     0x7f2986156824 - std::panicking::begin_panic::{{closure}}::h0c614cca9e230479
  13:     0x7f298615660c - std::sys_common::backtrace::__rust_end_short_backtrace::hff9d60ca938ec3c8
  14:     0x7f29861567fc - std::panicking::begin_panic::hb3668001c5a6a0ef
  15:     0x7f2986230264 - rustc_mir_build::thir::pattern::_match::is_useful::h95c720a73d8bf73b
  16:     0x7f2986230b76 - rustc_mir_build::thir::pattern::_match::is_useful_specialized::h71f5594a54ec8177
  17:     0x7f29861b783a - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold::h14b05e93c28a55a4
  18:     0x7f298622ed59 - rustc_mir_build::thir::pattern::_match::is_useful::h95c720a73d8bf73b
  19:     0x7f2986230b76 - rustc_mir_build::thir::pattern::_match::is_useful_specialized::h71f5594a54ec8177
  20:     0x7f29861b783a - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold::h14b05e93c28a55a4
  21:     0x7f298622ed59 - rustc_mir_build::thir::pattern::_match::is_useful::h95c720a73d8bf73b
  22:     0x7f2986230b76 - rustc_mir_build::thir::pattern::_match::is_useful_specialized::h71f5594a54ec8177
  23:     0x7f29861b783a - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold::h14b05e93c28a55a4
  24:     0x7f298622ed59 - rustc_mir_build::thir::pattern::_match::is_useful::h95c720a73d8bf73b
  25:     0x7f298619feb8 - <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr::h60c1881345d01731
  26:     0x7f298619fb3c - <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr::h60c1881345d01731
  27:     0x7f298619f863 - rustc_mir_build::thir::pattern::check_match::check_match::h47702a89d8c88d5e
  28:     0x7f2985211c59 - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::check_match>::compute::h4ed43a3b950e3311
  29:     0x7f298517b244 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h459d800570541209
  30:     0x7f2985214c34 - rustc_data_structures::stack::ensure_sufficient_stack::hc0f1c482e59b15b2
  31:     0x7f29851458ef - rustc_query_system::query::plumbing::get_query_impl::h0c4ce7dae7289273
  32:     0x7f2985160677 - rustc_query_system::query::plumbing::ensure_query_impl::h9e744db5e44fd749
  33:     0x7f29851e78ab - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h8768d28d97d9b4c6
  34:     0x7f298520ee18 - rustc_interface::passes::analysis::hf34e525a2b31f7c0
  35:     0x7f2984f48d7b - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute::h23cb841af51f0584
  36:     0x7f2984fe7057 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h7178e9f1ddee6943
  37:     0x7f2984fe23df - rustc_middle::ty::query::plumbing::<impl rustc_query_system::query::QueryContext for rustc_middle::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}::{{closure}}::h4025824db8759ceb
  38:     0x7f2984f73f10 - rustc_query_system::query::plumbing::get_query_impl::h29928f066778d4c4
  39:     0x7f2984f4b12b - rustc_interface::passes::QueryContext::enter::h0705343d8e4a95be
  40:     0x7f2984fe44a2 - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter::h6eb07f11eba80bd6
  41:     0x7f2984fab2a2 - rustc_span::with_source_map::h3820d41b3a036910
  42:     0x7f2984fe59c2 - rustc_interface::interface::create_compiler_and_run::h08ad396f4e7721e0
  43:     0x7f2984fcdd5a - scoped_tls::ScopedKey<T>::set::hd425a36fbe06985e
  44:     0x7f2984fea9b5 - std::sys_common::backtrace::__rust_begin_short_backtrace::h23925571219dbc63
  45:     0x7f2984f683ce - core::ops::function::FnOnce::call_once{{vtable.shim}}::h43b61b9af30b3669
  46:     0x7f298471021a - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h88864ee659ddb161
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/alloc/src/boxed.rs:1042:9
  47:     0x7f298471021a - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h9f03cbdc7ab33431
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/alloc/src/boxed.rs:1042:9
  48:     0x7f298471021a - std::sys::unix::thread::Thread::new::thread_start::h5979dd5560a11418
                               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sys/unix/thread.rs:89:17
  49:     0x7f29845f53e9 - start_thread
  50:     0x7f29843dd293 - __GI___clone
  51:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new

note: Clippy version: clippy 0.0.212 (ffa2e7a 2020-10-24)

query stack during panic:
#0 [check_match] match-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
warning: 1 warning emitted
  ```
  
  </p>
</details>

**Comment by ebroto** [comment]
This seems to be crashing during the `analysis` pass of rustc, no? (so, before Clippy runs its late passes)
